packages/miniflare/src/plugins/hyperdrive/hyperdrive-proxy.ts

```diff
@@ -0,0 +1,427 @@
+	0x00, 0x00, 0x00, 0x08, 0x04, 0xd2, 0x16, 0x2f,
+]);
+
+/**
+ * HyperdriveProxyController establishes TLS-enabled connections between workerd
+ * and external Postgres/MySQL databases. Supports PostgreSQL sslmode options
+ * ('require', 'prefer', 'disable') by proxying each Hyperdrive binding through
+ * a randomly assigned local port.
+ */
+export class HyperdriveProxyController {
+	// Map hyperdrive binding name to proxy server
+	#servers = new Map<string, net.Server>();
+
+	/**
+	 * Creates a proxy server for a Hyperdrive binding.
+	 *
+	 * @param config - The configuration for the proxy server.
+	 * @returns A promise that resolves to the port number of the proxy server.
+	 */
+	async createProxyServer(config: HyperdriveProxyConfig): Promise<number> {
+		const { name, targetHost, targetPort, scheme, sslmode } = config;
+		const server = net.createServer((clientSocket) => {
+			this.#handleConnection(
+				clientSocket,
+				targetHost,
+				Number.parseInt(targetPort),
+				scheme,
+				sslmode
+			);
+		});
+		const port = await new Promise<number>((resolve) => {
+			server.listen(0, "localhost", () => {
+				const addr = server.address() as net.AddressInfo;
+				resolve(addr.port);
+			});
+		});
+		this.#servers.set(name, server);
+		return port;
+	}
+
+	/**
+	 * Handles a connection from workerd by hyperdrive magic hostname "randomUUID.hyperdrive.local"
+	 *
+	 * @param clientSocket - The client socket.
+	 * @param targetHost - The hostname of the target database.
+	 * @param targetPort - The port of the target database.
+	 * @param scheme - The scheme of the target database.
+	 * @param sslmode - The sslmode of the target database.
+	 */
+	async #handleConnection(
+		clientSocket: net.Socket,
+		targetHost: string,
+		targetPort: number,
+		scheme: string,
+		sslmode: string
+	) {
+		// Connect to real database
+		const dbSocket = net.connect({ host: targetHost, port: targetPort });
+		const sslmodeRequire = sslmode === "require";
+		const sslmodePrefer = sslmode === "prefer";
+		if (sslmodePrefer || sslmodeRequire) {
+			try {
+				if (scheme === schemes.postgres || scheme === schemes.postgresql) {
+					return await handlePostgresTlsConnection(
+						dbSocket,
+						clientSocket,
+						targetHost,
+						targetPort,
+						sslmodeRequire
+					);
+				} else if (scheme === schemes.mysql) {
+					return await handleMySQLTlsConnection(
+						dbSocket,
+						clientSocket,
+						targetHost,
+						targetPort,
+						sslmodeRequire
+					);
+				}
+			} catch (e) {
+				if (sslmodeRequire) {
+					// Write error to client so worker can read it
+					clientSocket.write(`${e}\n`);
+					clientSocket.end();
+					dbSocket.destroy();
+					return;
+				}
+			}
+		}
+		// Pipe plain tcp sockets
+		clientSocket.pipe(dbSocket);
+		dbSocket.pipe(clientSocket);
+	}
+
+	/** Disposes of the proxy servers when shutting down the worker.*/
+	async dispose(): Promise<void> {
+		await Promise.all(
```

Comment:
`all` will reject as soon as one of the Promise rejects.
`allSettled` will wait for all Promises to settle (either successfully or not).

```suggestion
		await Promise.allSettled(
```

Would there be any value in adding a debug log on error?